physical/raft/raft_autopilot.go

```diff
@@ -0,0 +1,718 @@
+		return time.Duration(0)
+	}
+	return time.Duration(*d)
+}
+
+func (d *ReadableDuration) MarshalJSON() ([]byte, error) {
+	return []byte(fmt.Sprintf(`"%s"`, d.Duration().String())), nil
+}
+
+func (d *ReadableDuration) UnmarshalJSON(raw []byte) (err error) {
+	if d == nil {
+		return fmt.Errorf("cannot unmarshal to nil pointer")
+	}
+
+	var dur time.Duration
+	str := string(raw)
+	if len(str) >= 2 && str[0] == '"' && str[len(str)-1] == '"' {
+		// quoted string
+		dur, err = time.ParseDuration(str[1 : len(str)-1])
+		if err != nil {
+			return err
+		}
+	} else {
+		// no quotes, not a string
+		v, err := strconv.ParseFloat(str, 64)
+		if err != nil {
+			return err
+		}
+		dur = time.Duration(v)
+	}
+
+	*d = ReadableDuration(dur)
+	return nil
+}
+
+func stringIDs(ids []raft.ServerID) []string {
+	out := make([]string, len(ids))
+	for i, id := range ids {
+		out[i] = string(id)
+	}
+	return out
+}
+
+func autopilotToAPIState(state *autopilot.State) *AutopilotState {
+	out := &AutopilotState{
+		Healthy:          state.Healthy,
+		FailureTolerance: state.FailureTolerance,
+		Leader:           string(state.Leader),
+		Voters:           stringIDs(state.Voters),
+		Servers:          make(map[string]*AutopilotServer),
+	}
+
+	for id, srv := range state.Servers {
+		out.Servers[string(id)] = autopilotToAPIServer(srv)
+	}
+
+	autopilotToAPIStateEnterprise(state, out)
+
+	return out
+}
+
+func autopilotToAPIServer(srv *autopilot.ServerState) *AutopilotServer {
+	apiSrv := &AutopilotServer{
+		ID:          string(srv.Server.ID),
+		Name:        srv.Server.Name,
+		Address:     string(srv.Server.Address),
+		NodeStatus:  string(srv.Server.NodeStatus),
+		LastContact: NewReadableDuration(srv.Stats.LastContact),
+		LastTerm:    srv.Stats.LastTerm,
+		LastIndex:   srv.Stats.LastIndex,
+		Healthy:     srv.Health.Healthy,
+		StableSince: srv.Health.StableSince,
+		Status:      string(srv.State),
+		Meta:        srv.Server.Meta,
+	}
+
+	autopilotToAPIServerEnterprise(srv, apiSrv)
+
+	return apiSrv
+}
+
+// GetAutopilotServerState retrieves raft cluster state from autopilot to
+// return over the API.
+func (b *RaftBackend) GetAutopilotServerState(ctx context.Context) (*AutopilotState, error) {
+	b.l.RLock()
+	defer b.l.RUnlock()
+
+	if b.raft == nil {
+		return nil, errors.New("raft storage is not initialized")
+	}
+
+	if b.autopilot == nil {
+		return nil, nil
+	}
+
+	apState := b.autopilot.GetState()
+	if apState == nil {
+		return nil, nil
+	}
+
+	state := autopilotToAPIState(apState)
+
+	apStatus, _ := b.autopilot.IsRunning()
+	state.ExecutionStatus = autopilotStatusToStatus(apStatus)
+
+	return state, nil
+}
+
+func (b *RaftBackend) DisableAutopilot() {
+	b.l.Lock()
+	b.disableAutopilot = true
+	b.l.Unlock()
+}
+
+// SetupAutopilot instantiates autopilot and tells if its okay to start it.
+func (b *RaftBackend) SetupAutopilot(ctx context.Context, storageConfig *AutopilotConfig) {
+	// TODO: Adding this lock is creating a dead lock. Figure it out.
```

Comment:
You might want to try using https://github.com/sasha-s/go-deadlock.

Reply:
I was curious so I gave it a look.  If you ever want to do this btw, just change the package of the mutex type from `sync` to `deadlock` (as with Core.stateLock) and run your test with the build tag `deadlock`.  That yields

```
POTENTIAL DEADLOCK: Duplicate locking, saw callers this locks in one goroutine:
current goroutine 282 lock &{{{1 0} 0 0 -1073741824 0}} 
all callers to this lock in the goroutine
2021-02-26T12:26:57.049-0500 [TRACE] core0.activity: scanned existing logs: out=[]
../../../physical/raft/raft_autopilot.go:658 raft.(*RaftBackend).SetupAutopilot { b.l.Lock() } <<<<<
../../raft.go:182 vault.(*Core).setupRaftActiveNode { raftBackend.SetupAutopilot(c.activeContext, autopilotConfig) }
../../core.go:2002 vault.standardUnsealStrategy.unseal { if err := c.setupRaftActiveNode(ctx); err != nil { }
../../core.go:2056 vault.(*Core).postUnseal { if err := unsealer.unseal(ctx, c.logger, c); err != nil { }
../../ha.go:565 vault.(*Core).waitForLeadership { err = c.postUnseal(activeCtx, activeCtxCancel, standardUnsealStrategy{}) }
../../ha.go:387 vault.(*Core).runStandby.func7 { c.waitForLeadership(newLeaderCh, manualStepDownCh, leaderStopCh) }
~/go/pkg/mod/github.com/oklog/run@v1.0.0/group.go:38 run.(*Group).Run.func1 { errors <- a.execute() }

../../../physical/raft/raft_autopilot.go:249 raft.(*Delegate).AutopilotConfig { d.l.RLock() } <<<<<
~/go/pkg/mod/github.com/hashicorp/raft-autopilot@v0.1.2/state.go:61 raft-autopilot.(*Autopilot).gatherNextStateInputs { config := a.delegate.AutopilotConfig() }
~/go/pkg/mod/github.com/hashicorp/raft-autopilot@v0.1.2/state.go:133 raft-autopilot.(*Autopilot).nextState { inputs, err := a.gatherNextStateInputs(ctx) }
~/go/pkg/mod/github.com/hashicorp/raft-autopilot@v0.1.2/state.go:333 raft-autopilot.(*Autopilot).updateState { newState, err := a.nextState(ctx) }
~/go/pkg/mod/github.com/hashicorp/raft-autopilot@v0.1.2/run.go:38 raft-autopilot.(*Autopilot).Start { a.updateState(ctx) }
../../../physical/raft/raft_autopilot.go:681 raft.(*RaftBackend).SetupAutopilot { b.autopilot.Start(ctx) }
../../raft.go:182 vault.(*Core).setupRaftActiveNode { raftBackend.SetupAutopilot(c.activeContext, autopilotConfig) }
../../core.go:2002 vault.standardUnsealStrategy.unseal { if err := c.setupRaftActiveNode(ctx); err != nil { }
../../core.go:2056 vault.(*Core).postUnseal { if err := unsealer.unseal(ctx, c.logger, c); err != nil { }
../../ha.go:565 vault.(*Core).waitForLeadership { err = c.postUnseal(activeCtx, activeCtxCancel, standardUnsealStrategy{}) }
../../ha.go:387 vault.(*Core).runStandby.func7 { c.waitForLeadership(newLeaderCh, manualStepDownCh, leaderStopCh) }
~/go/pkg/mod/github.com/oklog/run@v1.0.0/group.go:38 run.(*Group).Run.func1 { errors <- a.execute() }
```

So I think all you need to do is release the lock before you call b.autopilot.Start().

Reply:
Noted the library suggestion. The deadlock stopped showing up after the refactor and simplification.

Reply:
Fixed.